pkg/git/repository/gitlab/service_test.go

```diff
@@ -0,0 +1,219 @@
+
+	// when
+	service, err := gitlab.NewRepoServiceIfMatches()(source, git.NewOauthToken([]byte("some-token")))
+
+	// then
+	assert.NoError(t, err)
+	assert.NotNil(t, service)
+}
+
+func TestNewRepoServiceIfMatchesShouldNotFailWhenSsh(t *testing.T) {
+	// given
+	source := test.NewGitSource(test.WithURL("git@gitlab.com:" + repoIdentifier))
+
+	// when
+	service, err := gitlab.NewRepoServiceIfMatches()(source, git.NewOauthToken([]byte("some-token")))
+
+	// then
+	assert.NoError(t, err)
+	assert.NotNil(t, service)
+}
+
+func TestRepositoryServiceForWrongRepo(t *testing.T) {
+	// given
+	defer gock.OffAll()
+	usernamePassword := git.NewUsernamePassword("anonymous", "")
+	oauthToken := git.NewOauthToken([]byte("some-token"))
+	mockTokenCall(t)
+
+	for _, secret := range []git.Secret{usernamePassword, oauthToken} {
+		gock.New("https://gitlab.com").
+			Get(fmt.Sprintf("/api/v4/projects/%s/", repoIdentifier)).
+			Times(2).
+			Reply(404).
+			BodyString(notFound)
+		source := test.NewGitSource(test.WithURL(repoURL), test.WithRef("dev"))
+
+		// when
+		service, err := gitlab.NewRepoServiceIfMatches()(source, secret)
+
+		// then
+		require.NoError(t, err)
+
+		filesInRootDir, err := service.GetListOfFilesInRootDir()
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "Not Found")
+		require.Len(t, filesInRootDir, 0)
+
+		languageList, err := service.GetLanguageList()
+		require.Error(t, err)
+		assert.Contains(t, err.Error(), "Not Found")
+		require.Len(t, languageList, 0)
+	}
+}
+
+func TestRepositoryServiceForPrivateInstance(t *testing.T) {
+	// given
+	defer gock.OffAll()
+	oauthToken := git.NewOauthToken([]byte("some-token"))
+
+	for _, url := range []string{"https://gitlab.cee.redhat.com/" + repoIdentifier, "git@gitlab.cee.redhat.com:" + repoIdentifier} {
+
+		mockGLCalls(t, "https://gitlab.cee.redhat.com/", repoIdentifier, "master", "some-token",
+			test.S("pom.xml", "mvnw"), test.S("Java", "Go"))
+
+		source := test.NewGitSource(test.WithURL(url), test.WithFlavor("gitlab"))
+
+		// when
+		service, err := gitlab.NewRepoServiceIfMatches()(source, oauthToken)
+
+		// then
+		require.NoError(t, err)
+
+		filesInRootDir, err := service.GetListOfFilesInRootDir()
+		require.NoError(t, err)
+		require.Len(t, filesInRootDir, 2)
+		assert.Contains(t, filesInRootDir, "pom.xml")
+		assert.Contains(t, filesInRootDir, "mvnw")
+
+		languageList, err := service.GetLanguageList()
+		require.NoError(t, err)
+		require.Len(t, languageList, 2)
+		assert.Contains(t, languageList, "Java")
+		assert.Contains(t, languageList, "Go")
+	}
+}
+
+func mockTokenCall(t *testing.T) {
+	token := &oauth2.Token{
+		AccessToken: "some-token",
+		TokenType:   "bearer",
+	}
+	bytes, err := json.Marshal(token)
+	require.NoError(t, err)
+	gock.New("https://gitlab.com").
+		Post("/oauth/token").
+		Reply(200).
+		BodyString(string(bytes))
+}
+
+func mockGLCalls(t *testing.T, host, prjPath, branch, token string, files, langs test.SliceOfStrings) {
```

Comment:
Would be nice to have some real e2e test against a gitlab instance at some point... but mocks are enough for now.

Reply:
There are real tests targeting the Gitlab instance: https://github.com/redhat-developer/git-service/pull/9/files#diff-8d3248024bbd50149e89d06034a64f4b but they are ignored for now. The reason is that I wanted to avoid failed tests because of unavailable service or changed repository structure. Rate limit failures are most likely not possible in case of Gitlab.
I'll create a repo project inside the Gitlab using my account and write a test targeting this repo.

Reply:
added 0ad74ef